File: src/Test/ListOpsFFI.php
<?php
$exports['runListOpsFFI'] = function($limit) {
    $n = (int)$limit;
    
    $range = function($start) {
        return function($end) use ($start) {
            $go = function($curr, $acc) use (&$go, $start) {
                if ($curr < $start) return $acc;
                return $go($curr - 1, (object)["type" => "Cons", "value0" => $curr, "value1" => $acc]);
            };
            return $go($end, (object)["type" => "Nil"]);
        };
    };

    $filter = function($p) {
        return function($lst) use ($p) {
            $go = function($list, $acc) use (&$go, $p) {
                if ($list->type === "Nil") {
                    return $acc;
                }
                $x = $list->value0;
                $xs = $list->value1;
                if ($p($x)) {
                    return $go($xs, (object)["type" => "Cons", "value0" => $x, "value1" => $acc]);
                } else {
                    return $go($xs, $acc);
                }
            };
            return $go($lst, (object)["type" => "Nil"]);
        };
    };

    $foldl = function($f, $initial, $lst) {
        $go = function($list, $acc) use (&$go, $f) {
            if ($list->type === "Nil") return $acc;
            return $go($list->value1, $f($acc)($list->value0));
        };
        return $go($lst, $initial);
    };

    $rng = $range(1)($n);
    $filtered = $filter(function($x) { return $x % 2 === 0; })($rng);
    return $foldl(function($acc) { return function($x) use ($acc) { return $acc + $x; }; }, 0, $filtered);
};
return $exports;

File: src/Test/ListOpsFFI.rs
enum List<A> {
    Nil,
    Cons(A, Box<List<A>>),
}

fn range_list(start: i64, end: i64) -> Box<List<i64>> {
    let mut acc = Box::new(List::Nil);
    let mut curr = end;
    while curr >= start {
        acc = Box::new(List::Cons(curr, acc));
        curr -= 1;
    }
    acc
}

fn filter_evens(mut lst: &List<i64>) -> Box<List<i64>> {
    let mut acc = Box::new(List::Nil);
    while let List::Cons(x, xs) = lst {
        if x % 2 == 0 {
            acc = Box::new(List::Cons(*x, acc));
        }
        lst = xs;
    }
    acc
}

fn foldl<A, B, F>(f: F, mut acc: B, mut lst: &List<A>) -> B
where
    F: Fn(B, &A) -> B,
{
    while let List::Cons(x, xs) = lst {
        acc = f(acc, x);
        lst = xs;
    }
    acc
}

pub fn Test_ListOpsFFI_runListOpsFFI(limit: i64) -> i64 {
    let lst = range_list(1, limit);
    let filtered = filter_evens(&lst);
    foldl(|a, b| a + *b, 0, &filtered)
}
